---
title: "Dev"
output: html_document
date: "2025-02-12"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## BSc project

This is a function to generate the population:

```{r}
population <- function(params) {
  #' Generate an initial population
  #'
  #' Generate the 0th generation of a population.
  #' @param params Path to JSON config file with simulation parameters.

  # Retrieve simulation parameters from configuration file
  config <- jsonlite::fromJSON(params)
 
  # Extract relevant parameters 
  n <- config$n_pop
  p <- config$n_gene
  q <- config$n_pheno

  # Simulate sex of individuals, minor allele frequencies and genotypes
  sex <- sample(rep(c(0, 1), n / 2))
  maf <- 0.5
  
  gene <- matrix(rbinom(n * p, 2, rep(maf, each = n)), nrow = n, ncol = p)

  names(maf) <- paste0("rs", 1:p)
  colnames(gene) <- paste0("rs", 1:p)

  # Generate phenotypes, for now model as function of genotype
  snps_causal <- config$snp_causal
  phi <- matrix(nrow = n, ncol = 0)
  gam_causal <- list()
  
  for (pheno in names(snps_causal)) {
    snp_pheno <- snps_causal[[pheno]]
    gene_snp <- gene[, snp_pheno]
    gam_snp <- matrix(2)# @TODO: Make into variable
    
    names(gam_snp) <- paste0("rs", snp_pheno)
    gam_causal[[pheno]] <- gam_snp
    
    phenos <- gene_snp %*% gam_snp + rnorm(n)
    colnames(phenos) <- pheno
    
    phi <- cbind(phi, phenos)
  }
  
  pop <- data.frame(sex, gene, phi)
  
  attr(pop, "maf_genes") <- maf
  attr(pop, "gam_causal") <- gam_causal
  
  return(pop)
}
```

Here are $\psi: \mathcal S \to \mathbf R$ and $\Delta \psi$ as discussed.

```{r}
dpsi <- function(sol, swap) norm(sol[swap, "phi1"] - sol[rev(swap), "phi1"], type = "2")^2
psi <- function(sol) norm(sol$phi1 - sol$f_phi1, type = "2")^2
```

Here we implement the `sim_matching` function to generate a spouse matching between the male and female subpopulations. Currently, we are only using a single trait $\phi_1$ to select on.

```{r}
sim_matching <- function(pop, iter = 10, kappa0 = 1, greedy = TRUE, eval = FALSE, progress = FALSE) {
  # Size of male and female subpopulations
  n <- nrow(pop) / 2
  T_int <- floor(seq(from = 1, to = iter, length.out = 10))
  
  # Initial sample -- change this to greedy
  males <- pop[pop$sex == 0, ]
  females <- pop[pop$sex == 1, ]
  
  if (greedy) {
    males <- males[order(males$phi1), ]
    females <- females[order(females$phi1), ]
  }
  
  rownames(males) <- rownames(females) <- 1:n
  names(females) <- paste0("f_", colnames(females))
  
  sol <- cbind(males, females)
  exch_col <- grepl("f_", names(sol))
  
  swaps <- replicate(iter, sample(n, size = 2), simplify = FALSE)
  kappa <- kappa0 / log(1 + 0.01 * 1:iter)
  
  if (eval) {
    psi_eval <- numeric(iter)
    dpsi_eval <- numeric(iter)
    rho_eval <- numeric(iter)
    psi_eval[1] <- psi(sol)
    dpsi_eval[1] <- 0
    rho_eval[1] <- 1
  }
  
  # Metropolis step
  for (s in 2:iter) {
    if (s %in% T_int && progress) print(s)
    
    swap <- swaps[[s]]
    dpsi_sol <- dpsi(sol, swap)
    rho_s <- min(1, exp(dpsi_sol / kappa))
    
    # Accept step
    if (runif(1) > rho_s) {
      sol[swap, exch_col] <- sol[rev(swap), exch_col]
    }
    
    if (eval) {
      psi_eval[s] <- psi(sol)
      dpsi_eval[s] <- dpsi_sol
      rho_eval[s] <- rho_s
    }
  }
    
  if (eval) {
    return(list(
      sol = sol,
      dpsi = dpsi_eval,
      psi = psi_eval,
      rho = rho_eval
    ))
  } else {
    return(list(
      sol = sol
    ))
  }
}

plot_eval <- function(eval) {
  par(mfrow = c(1, 3))
  plot(eval$psi, type = "l", xlab = "Iteration", ylab = expression(psi))
  plot(eval$dpsi, type = "l", xlab = "Iteration", ylab = expression("Delta psi"))
  hist(eval$rho)
}

test_matching <- sim_matching(pop, iter = 1000, eval = TRUE, greedy = FALSE)
plot_eval(test_matching)
```

Verification

```{r}
pop <- population("Github/AssocMating/config.json")
```
